### PR TITLE
[SPARK-28390][SQL][PYTHON][TESTS] Convert and port 'pgSQL/select_having.sql' into UDF test base

### PR DESCRIPTION
**Pull request by shivusondur**
## What changes were proposed in this pull request?
changed the test according to steps mentioned in SPARK-27921

<details>
<summary>difference comparing to select_having.sql</summary>
<p>

```diff
diff --git a/sql/core/src/test/resources/sql-tests/results/pgSQL/select_having.sql.out b/sql/core/src/test/resources/sql-tests/results/udf/pgSQL/udf-select_having.sql.out
index 02536eb..f731d11 100644
--- a/sql/core/src/test/resources/sql-tests/results/pgSQL/select_having.sql.out
+++ b/sql/core/src/test/resources/sql-tests/results/udf/pgSQL/udf-select_having.sql.out
@@ -91,54 +91,54 @@ struct<>
 
 
 -- !query 11
-SELECT b, c FROM test_having
-	GROUP BY b, c HAVING count(*) = 1 ORDER BY b, c
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(count(*)) = 1 ORDER BY udf(b), udf(c)
 -- !query 11 schema
-struct<b:int,c:string>
+struct<CAST(udf(cast(b as string)) AS INT):int,CAST(udf(cast(c as string)) AS STRING):string>
 -- !query 11 output
 1	XXXX
 3	bbbb
 
 
 -- !query 12
-SELECT b, c FROM test_having
-	GROUP BY b, c HAVING b = 3 ORDER BY b, c
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(b) = 3 ORDER BY udf(b), udf(c)
 -- !query 12 schema
-struct<b:int,c:string>
+struct<CAST(udf(cast(b as string)) AS INT):int,CAST(udf(cast(c as string)) AS STRING):string>
 -- !query 12 output
 3	BBBB
 3	bbbb
 
 
 -- !query 13
-SELECT c, max(a) FROM test_having
-	GROUP BY c HAVING count(*) > 2 OR min(a) = max(a)
+SELECT udf(c), max(udf(a)) FROM test_having
+	GROUP BY c HAVING udf(count(*)) > 2 OR udf(min(a)) = udf(max(a))
 	ORDER BY c
 -- !query 13 schema
-struct<c:string,max(a):int>
+struct<CAST(udf(cast(c as string)) AS STRING):string,max(CAST(udf(cast(a as string)) AS INT)):int>
 -- !query 13 output
 XXXX	0
 bbbb	5
 
 
 -- !query 14
-SELECT min(a), max(a) FROM test_having HAVING min(a) = max(a)
+SELECT udf(udf(min(udf(a)))), udf(udf(max(udf(a)))) FROM test_having HAVING udf(udf(min(udf(a)))) = udf(udf(max(udf(a))))
 -- !query 14 schema
-struct<min(a):int,max(a):int>
+struct<CAST(udf(cast(cast(udf(cast(min(cast(udf(cast(a as string)) as int)) as string)) as int) as string)) AS INT):int,CAST(udf(cast(cast(udf(cast(max(cast(udf(cast(a as string)) as int)) as string)) as int) as string)) AS INT):int>
 -- !query 14 output
 
 
 
 -- !query 15
-SELECT min(a), max(a) FROM test_having HAVING min(a) < max(a)
+SELECT udf(min(udf(a))), udf(udf(max(a))) FROM test_having HAVING udf(min(a)) < udf(max(udf(a)))
 -- !query 15 schema
-struct<min(a):int,max(a):int>
+struct<CAST(udf(cast(min(cast(udf(cast(a as string)) as int)) as string)) AS INT):int,CAST(udf(cast(cast(udf(cast(max(a) as string)) as int) as string)) AS INT):int>
 -- !query 15 output
 0	9
 
 
 -- !query 16
-SELECT a FROM test_having HAVING min(a) < max(a)
+SELECT udf(a) FROM test_having HAVING udf(min(a)) < udf(max(a))
 -- !query 16 schema
 struct<>
 -- !query 16 output
@@ -147,16 +147,16 @@ grouping expressions sequence is empty, and 'default.test_having.`a`' is not an
 
 
 -- !query 17
-SELECT 1 AS one FROM test_having HAVING a > 1
+SELECT 1 AS one FROM test_having HAVING udf(a) > 1
 -- !query 17 schema
 struct<>
 -- !query 17 output
 org.apache.spark.sql.AnalysisException
-cannot resolve '`a`' given input columns: [one]; line 1 pos 40
+cannot resolve '`a`' given input columns: [one]; line 1 pos 44
 
 
 -- !query 18
-SELECT 1 AS one FROM test_having HAVING 1 > 2
+SELECT 1 AS one FROM test_having HAVING udf(udf(1) > udf(2))
 -- !query 18 schema
 struct<one:int>
 -- !query 18 output
@@ -164,7 +164,7 @@ struct<one:int>
 
 
 -- !query 19
-SELECT 1 AS one FROM test_having HAVING 1 < 2
+SELECT 1 AS one FROM test_having HAVING udf(udf(1) < udf(2))
 -- !query 19 schema
 struct<one:int>
 -- !query 19 output
@@ -172,7 +172,7 @@ struct<one:int>
 
 
 -- !query 20
-SELECT 1 AS one FROM test_having WHERE 1/a = 1 HAVING 1 < 2
+SELECT 1 AS one FROM test_having WHERE 1/udf(a) = 1 HAVING 1 < 2
 -- !query 20 schema
 struct<one:int>
 -- !query 20 output
```
</p>
</details>

## How was this patch tested?
by:

```bash
sudo SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite -- -z udf/pgSQL/udf-select_having.sql"
```

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Thanks for working on this. Let's wait for https://github.com/apache/spark/pull/25130. We need to fix UDF testing base itself

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) = udf(max(a));
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) < udf(max(a));
+
+-- errors: ungrouped column references
+SELECT udf(a) FROM test_having HAVING udf(min(a)) < udf(max(a));
+SELECT 1 AS one FROM test_having HAVING udf(a) > 1;
+
+-- the really degenerate case: need not scan table at all
+SELECT 1 AS one FROM test_having HAVING 1 > 2;
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: After merging the PR I pointed out, let's try `HAVING udf(1 > 2)` combination as well.

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 2]
```diff
@@ -0,0 +1,57 @@
+--
+-- Portions Copyright (c) 1996-2019, PostgreSQL Global Development Group
+--
+--
+-- This test file was converted from inputs/pgSQL/select_having.sql
+--
+-- SELECT_HAVING
+-- https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/select_having.sql
+--
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: I would comment this:

```
--
-- Portions Copyright (c) 1996-2019, PostgreSQL Global Development Group
--
--
-- SELECT_HAVING
-- https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/select_having.sql
--
-- This test file was converted from inputs/pgSQL/select_having.sql
```

to keep the original comment intact.

**Comment by HyukjinKwon** [comment]
Can you post the diff in the PR description as guided in the parent JIRA?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 2 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Corrected

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 1 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Done

**Comment by shivusondur** [comment]
> Can you post the diff in the PR description as guided in the parent JIRA?

Done

**Comment by HyukjinKwon** [comment]
PR description isn't exactly guided. It should be foldable with <details> html tag.

**Comment by shivusondur** [comment]
> PR description isn't exactly guided. It should be foldable with
@HyukjinKwon 
Handled, Thanks for reminding me.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
#25130 is merged. Can you rebase and update it against the current master?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) = udf(max(a));
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) < udf(max(a));
+
+-- errors: ungrouped column references
+SELECT udf(a) FROM test_having HAVING udf(min(a)) < udf(max(a));
+SELECT 1 AS one FROM test_having HAVING udf(a) > 1;
+
+-- the really degenerate case: need not scan table at all
+SELECT 1 AS one FROM test_having HAVING udf(1 > 2);
+SELECT 1 AS one FROM test_having HAVING udf(1 < 2);
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: @shivusondur, can we try this combination too? `udf(udf(1) > udf(2))`

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+SELECT udf(c), max(udf(a)) FROM test_having
+	GROUP BY c HAVING udf(count(*)) > 2 OR udf(min(a)) = udf(max(a))
+	ORDER BY c;
+
+-- test degenerate cases involving HAVING without GROUP BY
+-- Per SQL spec, these should generate 0 or 1 row, even without aggregates
+
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) = udf(max(a));
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) < udf(max(a));
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: here too:

```
udf(udf(min(a)) < udf(max(a)))
```

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+INSERT INTO test_having VALUES (3, 3, 'BBBB', 'D');
+INSERT INTO test_having VALUES (4, 3, 'BBBB', 'e');
+INSERT INTO test_having VALUES (5, 3, 'bbbb', 'F');
+INSERT INTO test_having VALUES (6, 4, 'cccc', 'g');
+INSERT INTO test_having VALUES (7, 4, 'cccc', 'h');
+INSERT INTO test_having VALUES (8, 4, 'CCCC', 'I');
+INSERT INTO test_having VALUES (9, 4, 'CCCC', 'j');
+
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(count(*)) = 1 ORDER BY b, c;
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Could we try `GROUP BY udf(b), c` too?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+-- HAVING is effectively equivalent to WHERE in this case
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(b) = 3 ORDER BY b, c;
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: `ORDER BY udf(b), c` too

**Comment by HyukjinKwon** [comment]
Looks fine otherwise if the tests pass

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 5 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Not Done
getting below errors.
"org.apache.spark.sql.AnalysisException
expression 'default.test_having.`b`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;"

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 6 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Not done, Bellow is the error

"org.apache.spark.sql.AnalysisException
expression 'default.test_having.`b`' is neither present in the group by, nor is it an aggregate function. Add to group by or wrap in first() (or first_value) if you don't care which value you get.;"

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 4 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Handled

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 3 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Handled

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+-- 	ORDER BY lower(c);
+
+SELECT udf(c), max(udf(a)) FROM test_having
+	GROUP BY c HAVING udf(count(*)) > 2 OR udf(min(a)) = udf(max(a))
+	ORDER BY c;
+
+-- test degenerate cases involving HAVING without GROUP BY
+-- Per SQL spec, these should generate 0 or 1 row, even without aggregates
+
+SELECT udf(min(a)), udf(max(a)) FROM test_having HAVING udf(min(a)) = udf(max(a));
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Can we test `udf(min(udf(a)))` case too?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+INSERT INTO test_having VALUES (7, 4, 'cccc', 'h');
+INSERT INTO test_having VALUES (8, 4, 'CCCC', 'I');
+INSERT INTO test_having VALUES (9, 4, 'CCCC', 'j');
+
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(count(*)) = 1 ORDER BY b, c;
+
+-- HAVING is effectively equivalent to WHERE in this case
+SELECT udf(b), udf(c) FROM test_having
+	GROUP BY b, c HAVING udf(b) = 3 ORDER BY b, c;
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Hm, @shivusondur, seems working if we put udf in `ORDER BY`

```python
from pyspark.sql.functions import pandas_udf, PandasUDFType

@pandas_udf("int", PandasUDFType.SCALAR)
def noop(x):
    return x

spark.udf.register("udf", noop)

spark.range(10).selectExpr("id as b", "id as c").createOrReplaceTempView("test_having")
sql("""SELECT udf(b), udf(c) FROM test_having
	GROUP BY b, c HAVING udf(count(*)) = 1 ORDER BY udf(b), c""").show()
```

```
+------+------+
|udf(b)|udf(c)|
+------+------+
|     0|     0|
|     1|     1|
|     2|     2|
|     3|     3|
|     4|     4|
|     5|     5|
|     6|     6|
|     7|     7|
|     8|     8|
|     9|     9|
+------+------+
```

can you double check this?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 7 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Handled, i added all combination

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 8 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Handled, i added all the combination

**Comment by HyukjinKwon** [comment]
@shivusondur, is the diff correct? Looks weird:

```diff
diff --git a/sql/core/src/test/resources/sql-tests/results/pgSQL/select_having.sql.out b/sql/core/src/test/resources/sql-tests/results/udf/pgSQL/udf-select_having.sql.out
index 02536eb..96ccd15 100644
--- a/sql/core/src/test/resources/sql-tests/results/pgSQL/select_having.sql.out
+++ b/sql/core/src/test/resources/sql-tests/results/udf/pgSQL/udf-select_having.sql.out
@@ -1,5 +1,5 @@
 -- Automatically generated by SQLQueryTestSuite
--- Number of queries: 22
+-- Number of queries: 42
```

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 9]
```diff
@@ -0,0 +1,84 @@
+--
+-- Portions Copyright (c) 1996-2019, PostgreSQL Global Development Group
+--
+--
+-- SELECT_HAVING
+-- https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/select_having.sql
+--
+-- This test file was converted from inputs/pgSQL/select_having.sql
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Can we add a comment like:

```
-- TODO: We should add UDFs in GROUP BY clause when [SPARK-28445] is resolved.
```

Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: > DIFF is correct. I added more combination of queries with udf() addition, so queries increased.
> Added the "-- TODO: We should add UDFs in GROUP BY clause when [SPARK-28445] is resolved." comment

Comment (HyukjinKwon) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: @shivusondur, let's avoid to add more combinations. `select_having.sql` file is being matched and ported from https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/select_having.sql

It's easier to review and check the diff if we match.

If we need, we can add separate test cases later in a different place but that's not what this JIRA targets.

Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Thanks for updating and addressing comments. Looks fine. Let me take a final look again before merging it.

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 10]
```diff
@@ -0,0 +1,58 @@
+--
+-- Portions Copyright (c) 1996-2019, PostgreSQL Global Development Group
+--
+--
+-- SELECT_HAVING
+-- https://github.com/postgres/postgres/blob/REL_12_BETA2/src/test/regress/sql/select_having.sql
+--
+-- This test file was converted from inputs/pgSQL/select_having.sql
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Can we add this todo on the top:
```
-- TODO: We should add UDFs in GROUP BY clause when [SPARK-28445] is resolved.
```

and remove same lines below?

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+SELECT udf(c), max(udf(a)) FROM test_having
+	GROUP BY c HAVING udf(count(*)) > 2 OR udf(min(a)) = udf(max(a))
+	ORDER BY c;
+
+-- test degenerate cases involving HAVING without GROUP BY
+-- Per SQL spec, these should generate 0 or 1 row, even without aggregates
+
+SELECT udf(udf(min(udf(a)))), udf(udf(max(udf(a)))) FROM test_having HAVING udf(udf(min(udf(a)))) = udf(udf(max(udf(a))));
+SELECT udf(udf(min(udf(a)))), udf(udf(max(udf(a)))) FROM test_having HAVING udf(udf(min(udf(a)))) < udf(udf(max(udf(a))));
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Can we use different UDF combination comparing to the one right above?

**Review by HyukjinKwon — APPROVED** [review]
Looks good otherwise

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 11 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: updated like below

SELECT udf(udf(min(udf(a)))), udf(udf(max(udf(a)))) FROM test_having HAVING udf(udf(min(udf(a)))) = udf(udf(max(udf(a))));
SELECT udf(min(udf(a))), udf(udf(max(a))) FROM test_having HAVING udf(min(a)) < udf(max(udf(a)));

File: sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql (hunk 10 above)
Comment (shivusondur) [reply sql/core/src/test/resources/sql-tests/inputs/udf/pgSQL/udf-select_having.sql]: Done

**Comment by shivusondur** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**